File: frontend/src/pages/backlog/config/attachmentUpload.ts
/**
 * チケット添付で受け入れる Content-Type の許可リスト。
 *
 * `backend/internal/domain/attachment_upload.go` の `AcceptedAttachmentContentTypes` と
 * 1 対 1 で揃える（サーバー側の許可リストの単一ソースをそのまま写す。増減したら両方直す）。
 * `<input accept>` と選択直後のクライアント側早期弾きの両方でこの定義を使う。
 */
export const ACCEPTED_ATTACHMENT_CONTENT_TYPES = [
  'image/png',
  'image/jpeg',
  'image/gif',
  'image/webp',
  'application/pdf',
  'text/plain',
  'text/csv',
  'application/zip',
  'application/msword',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/vnd.ms-excel',
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
  'application/vnd.ms-powerpoint',
  'application/vnd.openxmlformats-officedocument.presentationml.presentation',
] as const;

/** `<input type="file" accept>` に渡す文字列（カンマ区切り）。 */
export const ACCEPTED_ATTACHMENT_ACCEPT_ATTR = ACCEPTED_ATTACHMENT_CONTENT_TYPES.join(',');

/** file.type が受け入れ対象の添付 MIME かを返す。 */
export function isAcceptedAttachmentContentType(type: string): boolean {
  return (ACCEPTED_ATTACHMENT_CONTENT_TYPES as readonly string[]).includes(type);
}

/**
 * 添付 1 件あたりの上限バイト数（25 MiB）。backend の `MaxAttachmentUploadBytes` と一致させる。
 *
 * サーバー側の判定を待たず、選んだ直後にクライアント側でも早期に弾く（無駄なアップロード
 * 開始と、待たされた末の失敗表示を避けるため。imageUpload.ts と同じ考え方）。
 */
export const MAX_ATTACHMENT_UPLOAD_BYTES = 25 * 1024 * 1024;

File: backend/internal/domain/attachment_upload.go
package domain

import "errors"

// AcceptedAttachmentContentTypes はチケット添付で受け付ける Content-Type の許可リスト。
// ここに無い型は ValidateAttachmentUpload が ErrUnsupportedAttachmentContentType で弾く
// （image_upload.go の ValidateImageUpload と同じ考え方 — 検証しないまま presign を焼くと
// 上限の無い PUT を許すのと同じ穴になる）。
//
// 添付は画像専用の rich-text アップロードと違い、文書・アーカイブも受け付ける。
// 実行可能形式（exe・sh 等）・マークアップ（html・svg。ブラウザでそのまま開かれると
// 保存型 XSS の経路になる）は許可リストに入れない。
var AcceptedAttachmentContentTypes = map[string]bool{
	"image/png":          true,
	"image/jpeg":         true,
	"image/gif":          true,
	"image/webp":         true,
	"application/pdf":    true,
	"text/plain":         true,
	"text/csv":           true,
	"application/zip":    true,
	"application/msword": true,
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document": true,
	"application/vnd.ms-excel": true,
	"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":         true,
	"application/vnd.ms-powerpoint":                                             true,
	"application/vnd.openxmlformats-officedocument.presentationml.presentation": true,
}

// MaxAttachmentUploadBytes は添付 1 件あたりのサイズ上限（25 MiB）。文書・アーカイブは
// 画像より大きくなりやすいため、rich-text 画像の上限（10 MiB）より緩めに取る。
const MaxAttachmentUploadBytes = 25 * 1024 * 1024

var (
	ErrUnsupportedAttachmentContentType = errors.New("unsupported attachment content type")
	ErrAttachmentTooLarge               = errors.New("attachment is too large")
)

// ValidateAttachmentUpload は添付アップロード要求（Content-Type とサイズ）が保存してよい
// 形かを見る。presign の前に必ず通す入口（ValidateImageUpload と同じ役割）。
func ValidateAttachmentUpload(contentType string, size int64) error {
	if !AcceptedAttachmentContentTypes[contentType] {
		return ErrUnsupportedAttachmentContentType
	}
	if size <= 0 || size > MaxAttachmentUploadBytes {
		return ErrAttachmentTooLarge
	}
	return nil
}
